src/js/utils/deviceFeaturesUtils.js

```diff
@@ -106,3 +106,30 @@ export const getUnavailableCapabilities = (
     });
     return list;
 };
+
+/**
+ * There is an inconsistency in representation of device feature revision attribute.
+ * - T1 uses UTF-8 representation. (df0963ec48f01f3d07ffca556e21ff0070cab099)
+ * - T2 uses hex raw bytes representation. (6466303936336563)
+ * To avoid being model specific, in case the inconsistency is fixed, it is required to reliably detect what encoding is used.
```

Comment:
I'm really struggling to understand all this encoding marshalling. Please correct me if I'm wrong.

Git identifies revisions by their SHA-1 hash expressed with the standard *[hexadecimal](https://www.wikiwand.com/en/Hexadecimal)* notation -- for example `5ef3efa42638d618e024c3a8f9197ffb03e243b8` for [this commit](https://github.com/trezor/connect/commit/5ef3efa42638d618e024c3a8f9197ffb03e243b8). And that's what we want to ultimately get here.

When a revisions is obtained from a TT device we get the hexadecimal representation of the hash as JS string directly.

When a revisions is obtained from T1 device, however, what we get isn't this hexadecimal representation but a hexadecimal representation of the bytes that form an utf8 sequence containing the original hexadecimal representation.

What we want here is to automatically distinguish these two cases and convert the kinda double-encoded variant to the simple one. Because commit sha only contains alfanumeric characters its utf8 representation encoded as hexadecimal contains only numbers[1]. So we first look if the incoming string contains a letter. If it does, it must be the simple variant and we can return it.

If the incoming string consists of numbers only it is very likely[2] the double-encoded variant hence we decode it (`Buffer.from(<incoming>` 'hex')`) and then put it to an utf8 string (`.toString('utf8')`) which now contains the simple *hexadecimal representation of the hash*.

Before we return this decoded variant we also check that it decoded to a valid hexadecimal. If it did, we return it. If it didn't it might mean that the incoming string was in fact the simple variant or that the data was corrupted.

There's still a small chance that a simple variant would be misclassified as double-encoded if 1) it consisted of numbers only 2) those numbers, read as a hexadecimal, would encode a valid utf8 sequence which would consist of alfa-numeric characters only. I won't try to calculate probability of that. .)

---

The comments in this PR are right, but also confusing because on first sight, they're contradictory. They give `df0963ec48f01f3d07ffca556e21ff0070cab099` as example of "utf8" encoding in opposition to `6466303936336563` a "[raw] hex" encoding. However most people, when shown `df0963ec48f01f3d07ffca556e21ff0070cab099`, would say it's a hexadecimal. I'm not really sure how to make them clearer. ./

Maybe it's just me. But it took me quite big effort to understand what is going on here.


[1] `0-9` maps to `30-39`, `a-f` maps to `61-66` and `A-Z` to `41-46`.
[2] Probability of getting all-numbers sha-1 hash is `(10/16)^40` or roughly `6.8e-9`.

Reply:
Yes, this is exactly what it means 👍  However, I have difficulties with how to express it clearly by only few words. I will give it another chance tomorrow.

Reply:
We could also link https://github.com/trezor/trezor-firmware/issues/1620 for more details.